File: Day3/LifeSupportRating.java
import java.util.*;
import jdk.jshell.spi.ExecutionControl.ResolutionException;
import java.io.File;
import java.io.FileNotFoundException;

public class LifeSupportRating {
    public static void main(String[] args) throws FileNotFoundException {
        Scanner sc = new Scanner(new File("input.txt"));
        ArrayList oxygen = new ArrayList<String>();
        ArrayList co2 = new ArrayList<String>();
        while(sc.hasNextLine()){
            String temp = sc.nextLine();
            if(temp.charAt(0) == '1'){
                oxygen.add(temp);
            }else{
                co2.add(temp);
            }
        }
        if(!(oxygen.size() > co2.size())){
            ArrayList temp = oxygen;
            oxygen = co2;
            co2 = temp;
        }
        System.out.println(value(oxygen, true)* value(co2, false));
    }

    private static int value(ArrayList<String> input, Boolean mostCommon){
        int pos = 1;
        ArrayList listWithOne = new ArrayList<String>();
        ArrayList listWithZero = new ArrayList<String>();
        while(input.size() != 1){
            for (String line : input) {
                if(line.charAt(pos) == '1'){
                    listWithOne.add(line);
                }
                else{
                    listWithZero.add(line);
                }
            }
            input.clear();
            if(mostCommon){
                if(listWithOne.size() >= listWithZero.size()){
                    input.addAll(listWithOne);
                }
                else{
                    input.addAll(listWithZero);
                }
            }else{
                if(listWithOne.size() < listWithZero.size()){
                    input.addAll(listWithOne);
                }
                else{
                    input.addAll(listWithZero);
                }
            }
            listWithOne.clear();
            listWithZero.clear();
            pos++;
        }
        int result = 0;
        int binaryPos = 1; 
        for(int i = 11; i >= 0; i--){
            if(input.get(0).charAt(i) == '1'){
                result += binaryPos;
            }
            binaryPos = binaryPos * 2;
        }
        return result;
    }
}

File: 2021/python/Day03/LifeSupportRating.py
def value(input, mostCommon):
    pos = 1
    listOne.clear()
    listZero.clear()
    while(len(input) != 1):
        for line in input:
            if(line[pos] == "1"):
                listOne.append(line)
            else:
                listZero.append(line)
        input.clear()
        if(mostCommon):
            if(len(listOne) >= len(listZero)):
                input.extend(listOne)
            else:
                input.extend(listZero)
        else:
            if(len(listOne) < len(listZero)):
                input.extend(listOne)
            else:
                input.extend(listZero)
        listOne.clear()
        listZero.clear()
        pos += 1
    return int(input[0], 2)
        
file = open('input.txt', 'r')
listOne = []
listZero = []
for line in file:
    if(line[0] == "1"):
        listOne.append(line)
    else:
        listZero.append(line)
oxygen = []
co2 = []
if(len(listOne) > len(listZero)):
    oxygen.extend(listOne)
    co2.extend(listZero)
else:
    oxygen.extend(listZero)
    co2.extend(listOne)
print(value(oxygen, True) * value(co2, False))